MOVIE RECOMMENDATION

Cosine similarity is the cosine of the angle between two n-dimensional vectors in an n-dimensional space. It is the dot product of the two vectors divided by the product of the two vectors' lengths (or magnitudes).

Example

In [404]:
from sklearn.metrics.pairwise import cosine_similarity

In [405]:
vec1 = [1,1,0,1,1]

In [406]:
vec2 = [0,1,0,1,1]

In [407]:
print(cosine_similarity([vec1, vec2]))

[[1.        0.8660254]
 [0.8660254 1.       ]]


Import Libraries

In [408]:
import numpy as np

In [409]:
import pandas as pd

In [410]:
import difflib

In [411]:
from sklearn.feature_extraction.text import TfidfVectorizer

# TfidfVectorizer - This is used to convert text data into numerical values

In [412]:
from sklearn.metrics.pairwise import cosine_similarity

In [413]:
import os

Check the current working directory

In [414]:
display (os.getcwd())

'C:\\Users\\Admin\\Desktop\\Movie recommendation system'

Change the current working directory and read data

In [415]:
os.chdir('C:\\Users\\Admin\\Desktop\\Movie recommendation system')

In [416]:
movies_data =pd.read_csv('movies.csv')

In [417]:
movies_data.head()

index     budget                                    genres  \
0      0  237000000  Action Adventure Fantasy Science Fiction   
1      1  300000000                  Adventure Fantasy Action   
2      2  245000000                    Action Adventure Crime   
3      3  250000000               Action Crime Drama Thriller   
4      4  260000000          Action Adventure Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash future space war space colony so...                en   
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   162.0   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  Sam Worthington Zoe Saldana Sigourney Weaver S...   
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
0  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[5 rows x 24 columns]

Display the shape

In [418]:
display (movies_data.shape)

(4803, 24)

In [419]:
# Selecting the relevant features for recommendation

In [420]:
selected_features = ['genres','keywords','tagline','cast','director']

In [421]:
print(selected_features)

['genres', 'keywords', 'tagline', 'cast', 'director']


Display the info

In [422]:
display (movies_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   index                 4803 non-null   int64  
 1   budget                4803 non-null   int64  
 2   genres                4775 non-null   object 
 3   homepage              1712 non-null   object 
 4   id                    4803 non-null   int64  
 5   keywords              4391 non-null   object 
 6   original_language     4803 non-null   object 
 7   original_title        4803 non-null   object 
 8   overview              4800 non-null   object 
 9   popularity            4803 non-null   float64
 10  production_companies  4803 non-null   object 
 11  production_countries  4803 non-null   object 
 12  release_date          4802 non-null   object 
 13  revenue               4803 non-null   int64  
 14  runtime               4801 non-null   float64
 15  spoken_languages     

None

In [423]:
# Check for Null Values

In [424]:
display (movies_data.isna().sum())

index                      0
budget                     0
genres                    28
homepage                3091
id                         0
keywords                 412
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title                      0
vote_average               0
vote_count                 0
cast                      43
crew                       0
director                  30
dtype: int64

Display the selected columns

In [425]:
display (movies_data[selected_features].head())

genres  \
0  Action Adventure Fantasy Science Fiction   
1                  Adventure Fantasy Action   
2                    Action Adventure Crime   
3               Action Crime Drama Thriller   
4          Action Adventure Science Fiction   

                                            keywords  \
0  culture clash future space war space colony so...   
1  ocean drug abuse exotic island east india trad...   
2         spy based on novel secret agent sequel mi6   
3  dc comics crime fighter terrorist secret ident...   
4  based on novel mars medallion space travel pri...   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                                cast           director  
0  Sam Worthington Zoe Saldana Sigourney Weaver S...      James Cameron  
1  Johnny Depp Orlando Bloom Keira Knightley Stel...     Gore Verbinski  
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...         Sam Mendes  
3  Christian Bale Michael Caine Gary Oldman Anne ...  Christopher Nolan  
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...     Andrew Stanton

Check Null Values in selected columns

In [426]:
display (movies_data[selected_features].isna().sum())

genres       28
keywords    412
tagline     844
cast         43
director     30
dtype: int64

Replacing the null values with null string

In [427]:
for feature in selected_features:
    movies_data[feature] = movies_data[feature].fillna('')

In [428]:
movies_data[feature] = movies_data[feature].fillna('')

In [429]:
display (movies_data.head())

index     budget                                    genres  \
0      0  237000000  Action Adventure Fantasy Science Fiction   
1      1  300000000                  Adventure Fantasy Action   
2      2  245000000                    Action Adventure Crime   
3      3  250000000               Action Crime Drama Thriller   
4      4  260000000          Action Adventure Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash future space war space colony so...                en   
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   162.0   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  Sam Worthington Zoe Saldana Sigourney Weaver S...   
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
0  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[5 rows x 24 columns]

Combining all the 5 selected features

In [430]:
combined_features = movies_data['genres']+' '+movies_data['keywords']+' '+movies_data['tagline']+' '+movies_data['cast']+' '+movies_data['director']

display (combined_features)

Converting the text data to feature vectors

This is to find cosine similarity

# Vector shape is (4803, 17318). This is based on the number of distinct words. All the words will be converted to their equivalent numbers.

In [431]:
vectorizer = TfidfVectorizer()

In [432]:
feature_vectors = vectorizer.fit_transform(combined_features)

display (feature_vectors.shape)

print (feature_vectors)

Getting the similarity scores using cosine similarity

In [433]:
similarity = cosine_similarity(feature_vectors)

print  (similarity )

Print Shape Cosine Similarity

In [434]:
display(similarity.shape)

(4803, 4803)

Enter the movie name to get Similarity

In [435]:
movie_name = input(' Enter your favourite movie name : ')

Creating a list with all the movie names given in the dataset

In [436]:
list_of_all_titles = movies_data['title'].tolist()

In [437]:
print(list_of_all_titles)

['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre', 'The Dark Knight Rises', 'John Carter', 'Spider-Man 3', 'Tangled', 'Avengers: Age of Ultron', 'Harry Potter and the Half-Blood Prince', 'Batman v Superman: Dawn of Justice', 'Superman Returns', 'Quantum of Solace', "Pirates of the Caribbean: Dead Man's Chest", 'The Lone Ranger', 'Man of Steel', 'The Chronicles of Narnia: Prince Caspian', 'The Avengers', 'Pirates of the Caribbean: On Stranger Tides', 'Men in Black 3', 'The Hobbit: The Battle of the Five Armies', 'The Amazing Spider-Man', 'Robin Hood', 'The Hobbit: The Desolation of Smaug', 'The Golden Compass', 'King Kong', 'Titanic', 'Captain America: Civil War', 'Battleship', 'Jurassic World', 'Skyfall', 'Spider-Man 2', 'Iron Man 3', 'Alice in Wonderland', 'X-Men: The Last Stand', 'Monsters University', 'Transformers: Revenge of the Fallen', 'Transformers: Age of Extinction', 'Oz: The Great and Powerful', 'The Amazing Spider-Man 2', 'TRON: Legacy', 'Cars 2', 'Green Lant

Length of the List

len(list_of_all_titles)

In [438]:
Finding the close match for the movie name given by the user

SyntaxError: invalid syntax (1302414561.py, line 1)

In [ ]:
find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)

In [ ]:
print(find_close_match)

["Pirates of the Caribbean: At World's End", "Pirates of the Caribbean: Dead Man's Chest", 'Pirates of the Caribbean: On Stranger Tides']


Display the close match

In [ ]:
close_match = find_close_match[0]

In [ ]:
print(close_match)

Pirates of the Caribbean: At World's End


Finding the index of the movie with the title

In [ ]:
index_of_the_movie = movies_data[movies_data.title == close_match]['index'].values[0]

In [ ]:
print(index_of_the_movie)

1


In [ ]:
similarity_score = list(enumerate(similarity[index_of_the_movie]))

In [ ]:
print(similarity_score)

[(0, np.float64(0.07219486822992488)), (1, np.float64(1.0)), (2, np.float64(0.032814990806793745)), (3, np.float64(0.020741502866825225)), (4, np.float64(0.03305463079248627)), (5, np.float64(0.09894823392218864)), (6, np.float64(0.014378790821108607)), (7, np.float64(0.019781895260853298)), (8, np.float64(0.04056180271185781)), (9, np.float64(0.029948196927440135)), (10, np.float64(0.06512426368922566)), (11, np.float64(0.02758093144680318)), (12, np.float64(0.532542301489058)), (13, np.float64(0.16315140691538466)), (14, np.float64(0.057522797936129924)), (15, np.float64(0.023846306520609505)), (16, np.float64(0.018757441743698592)), (17, np.float64(0.08693929581992484)), (18, np.float64(0.017747962646584)), (19, np.float64(0.06159820776263954)), (20, np.float64(0.0878081242460847)), (21, np.float64(0.016447740747051343)), (22, np.float64(0.047375560765684784)), (23, np.float64(0.04439777185587065)), (24, np.float64(0.07038585257947676)), (25, np.float64(0.0)), (26, np.float64(0.0204

Length of Similarity Score

len(similarity_score)

Sorting the movies based on their similarity score

Display the index and similarity rating as a tuple

This list is sorted based on the similarity score

In [ ]:
sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True)

In [ ]:
print(sorted_similar_movies)

[(1, np.float64(1.0)), (199, np.float64(0.5466481253706555)), (12, np.float64(0.532542301489058)), (536, np.float64(0.2888182386430639)), (1652, np.float64(0.19164772894913176)), (384, np.float64(0.18874711394669405)), (13, np.float64(0.16315140691538466)), (129, np.float64(0.1587339625172756)), (3055, np.float64(0.15577364618957562)), (3033, np.float64(0.15459290728645841)), (329, np.float64(0.15057603842736803)), (178, np.float64(0.14725915063827227)), (27, np.float64(0.14584867834417367)), (2035, np.float64(0.14149462529605059)), (267, np.float64(0.14000715404465197)), (1183, np.float64(0.13987145541658166)), (326, np.float64(0.13846182473796614)), (1095, np.float64(0.13674871062299215)), (1890, np.float64(0.13658339038642311)), (2401, np.float64(0.13627388719755612)), (2040, np.float64(0.13546842154573413)), (2474, np.float64(0.13515170184139363)), (2254, np.float64(0.1350451114801658)), (1577, np.float64(0.1299215119380567)), (894, np.float64(0.128822479300327)), (1061, np.float64

Print the name of similar movies based on the index – Top 30

In [ ]:

print('Movies suggested for you : \n')

Movies suggested for you : 



In [ ]:
i = 1

In [ ]:
for movie in sorted_similar_movies:
    index = movie[0]
    title_from_index = movies_data[movies_data.index==index]['title'].values[0]
    # Skip the input movie itself
    if title_from_index.lower() == close_match.lower():
        continue
    if i <= 30:
        print(i, '.', title_from_index)
        i += 1

1 . Pirates of the Caribbean: The Curse of the Black Pearl
2 . Pirates of the Caribbean: Dead Man's Chest
3 . Anna and the King
4 . Dragonball Evolution
5 . Cast Away
6 . The Lone Ranger
7 . Thor
8 . Seeking a Friend for the End of the World
9 . The Corruptor
10 . The Lord of the Rings: The Return of the King
11 . Rango
12 . Battleship
13 . Our Kind of Traitor
14 . Kingdom of Heaven
15 . The Mexican
16 . Cinderella
17 . Curse of the Golden Flower
18 . Don Juan DeMarco
19 . City of Ghosts
20 . The Glass House
21 . Arn: The Knight Templar
22 . The Children of Huang Shi
23 . Without a Paddle
24 . Bulletproof Monk
25 . Love Actually
26 . Noah
27 . Crouching Tiger, Hidden Dragon
28 . Star Trek II: The Wrath of Khan
29 . From Hell
30 . The Replacement Killers


In [ ]:
index = movie[0]

In [ ]:
title_from_index = movies_data[movies_data.index==index]['title'].values[0]

In [ ]:
# Skip the input movie itself and print unique recommendations
if title_from_index.lower() == close_match.lower():
    continue
if i <= 30:
    print(i, '.', title_from_index)
    i += 1

SyntaxError: 'continue' not properly in loop (2097395763.py, line 3)

In [ ]:
print(i, '.',title_from_index)

31 . My Date with Drew


In [ ]:
i+=1

Consolidated code

In [ ]:
movie_name = input(' Enter your favourite movie name : ')

In [ ]:
list_of_all_titles = movies_data['title'].tolist()

In [ ]:
find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)

In [ ]:
close_match = find_close_match[0]

In [ ]:
index_of_the_movie = movies_data[movies_data.title == close_match]['index'].values[0]

In [ ]:
similarity_score = list(enumerate(similarity[index_of_the_movie]))

In [ ]:
sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True)

In [ ]:
print('Movies suggested for you : \n')

Movies suggested for you : 



In [440]:
i = 1

In [442]:
for movie in sorted_similar_movies:
    index = movie[0]
    # ...existing code...

In [443]:
title_from_index = movies_data[movies_data.index==index]['title'].values[0]

In [446]:
# Print top 30 recommended movies, skipping the input movie itself
for movie in sorted_similar_movies:
    index = movie[0]
    title_from_index = movies_data[movies_data.index==index]['title'].values[0]
    if title_from_index.lower() == close_match.lower():
        continue
    if i <= 30:
        print(i, '.', title_from_index)
        i += 1

1 . Pirates of the Caribbean: The Curse of the Black Pearl
2 . Pirates of the Caribbean: Dead Man's Chest
3 . Anna and the King
4 . Dragonball Evolution
5 . Cast Away
6 . The Lone Ranger
7 . Thor
8 . Seeking a Friend for the End of the World
9 . The Corruptor
10 . The Lord of the Rings: The Return of the King
11 . Rango
12 . Battleship
13 . Our Kind of Traitor
14 . Kingdom of Heaven
15 . The Mexican
16 . Cinderella
17 . Curse of the Golden Flower
18 . Don Juan DeMarco
19 . City of Ghosts
20 . The Glass House
21 . Arn: The Knight Templar
22 . The Children of Huang Shi
23 . Without a Paddle
24 . Bulletproof Monk
25 . Love Actually
26 . Noah
27 . Crouching Tiger, Hidden Dragon
28 . Star Trek II: The Wrath of Khan
29 . From Hell
30 . The Replacement Killers


In [447]:
print(i, '.',title_from_index)

31 . My Date with Drew


In [448]:
i+=1